### ISSUE TITLE
MA0062- Non-flags enums should not be marked with "FlagsAttribute"

### ISSUE THREAD

**Issue by AESSoftware** [meziantou/Meziantou.Analyzer#453]
**Context**
Rule identifier: _MA0062_
Target Framework: _.Any_
Language version: _Any_

**Describe the bug**
Over sensitive to special case value of "All = ~None"

**To Reproduce**
Add to Flags enum All = ~None,

````c#
Code to reproduce the issue
````

**Additional context**
Admittedly -1 (i.e. ~None === ~0) isn't a power of 2, but its binary representation is :-)

**Comment by meziantou** [comment]
The rule is correct as described "An enumeration marked with FlagsAttribute should only contain members whose values are powers of two or a combination of such members."

-1 is not a power of 2. A power of 2 only contains a single "1" in its binary representation. Is considered a power of 2 a number where  is true

The algorithm to check if a value is valid is:
- `0` => valid
- `x` when `x & (x - 1) == 0` (power of 2) => valid (match the implementation of .NET: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs,986b9ecef5c766cd,references)
- other values must be a combination of other values matching the previous rule

````c#
enum Sample
{
   None = 0b_00,
   A    = 0b_01,
   B    = 0b_10,
   C    = A | B, // 0b_00...011: ✔️ Not a power of 2, but a combination of other existing values
   All  = ~None, // 0b_11...111: ❌ Not a power of 2 and not a combination of other existing values
                 //         ^ this "1" doesn't match any other defined values
}
````

I'm not against adding an option (in `.editorconfig`) to detect this special-case and not report it. But, maybe not enable it by default.

I don't know what would be a good name for this option. Any idea?

**Comment by AESSoftware** [comment]
Hi,

 

Thanks for your reply. I can’t fault the logic of what’s valid and agree that a disabled-by-default option would be useful.

How about “Allow Lazy All Flags” for the name – pretty much summarizes the sentiment/intent

 

Adrian
